### PR TITLE
Initial commit

### PR DESCRIPTION
**Pull request by jean-edouard**
This is the cdrom daemon.

For now all it does is change the iso on a blktap-based virtual drive.

In the future, this daemon will replace atapi-pt-passthourgh and handle CDROM arbitration.

This won't be included in the build for now.

Signed-off-by: Jed lejosnej@ainfosec.com

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,2 +1 @@
-# cdrom-daemon
-The OpenXT daemon that handles guest CDROM operations
+This is the CDROM daemon.
```
Comment (rossphilipson) [inline README.md:1]: Nice readme

File: src/.gitignore [hunk 2]
```diff
@@ -0,0 +1,5 @@
+config.h
+config.h.in
+*.o
+stamp-h1
+cdrom-daemon
```
Comment (rossphilipson) [inline src/.gitignore]: Missing nl?

File: src/blktap.c [hunk 3]
```diff
@@ -0,0 +1,328 @@
...
+  unsigned int i, count;
+  int vdev, res = -1;
+
+  /* Find the CDROM virtual device for the given domid */
+  snprintf(xpath, sizeof(xpath), "/local/domain/0/backend/vbd/%d", domid);
+  devs = xs_directory(xs_handle, XBT_NULL, xpath, &count);
+  if (devs) {
+    for (i = 0; i < count; ++i) {
+      vdev = strtol(devs[i], NULL, 10);
+      type = xenstore_be_read(XBT_NULL, domid, vdev, "device-type");
```
Comment (rossphilipson) [inline src/blktap.c:44]: type leaked?

File: src/blktap.c [hunk 4]
```diff
@@ -0,0 +1,328 @@
...
+
+static int cdrom_tap_minor_of_vdev(int domid, int vdev)
+{
+  char *tmp;
+  int tap_minor = -1;
+
+  if (vdev < 0)
+    return -1;
+
+  tmp = xenstore_be_read(XBT_NULL, domid, vdev, "params");
```
Comment (rossphilipson) [inline src/blktap.c:69]: tmp leaked?

File: src/blktap.c [hunk 5]
```diff
@@ -0,0 +1,328 @@
...
+      tm = cdrom_tap_minor_of_vdev(domid, cdrom_vdev_of_domid(domid));
+      if (print)
+	printf("CDROM for domid %d: %d\n", domid, tm);
+      if (tm == tap_minor)
+	res++;
+      if (res >= max)
+	break;
+    }
+  }
+  free(domids);
```
Comment (rossphilipson) [inline src/blktap.c]: Move inside of if (domids) to be consistent

File: src/blktap.c (hunk 3 above)
Comment (jean-edouard) [reply src/blktap.c:44]: Hmm, I somehow managed to convince myself that the output of xs_read() didn't have to be free()ed! Fixing all the reads now...

Comment (rossphilipson) [reply src/blktap.c:44]: Ack, I figured something like that so I stopped reviewing until you fixed it.

**Comment by jean-edouard** [comment]
Pushed a new commit with fixes.
Also replaced some rogue tabs with spaces.

File: src/blktap.c [hunk 6]
```diff
@@ -0,0 +1,345 @@
...
+  tmp = list;
+  while(*tmp != NULL) {
+    if ((*tmp)->minor == tap_minor) {
+      tap = *tmp;
+      break;
+    }
+    tmp++;
+  }
+  if (tap == NULL)
+    return false;
```
Comment (rossphilipson) [inline src/blktap.c:224]: Missing a tap_ctl_free_list(list); here?

File: src/blktap.c [hunk 7]
```diff
@@ -0,0 +1,345 @@
...
+ *    In which case we create a new tapdev, destroy the blktap and recreate one
+ *     pointing to the new iso. (tapdev hotplug is explicitely not supported)
+ */
+bool blktap_change_iso(const char *path, int domid)
+{
+  int tap_minor, count, vdev, existing;
+  char tpath[256], phys[16], *new_tpath = NULL;
+
+  /* Get the virtual cdrom vdev and tap minor for the domid */
+  vdev = cdrom_vdev_of_domid(domid);
```
Comment (rossphilipson) [inline src/blktap.c:301]: Can't this fail to find something and return -1

File: src/main.c [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+  /* Main loop */
+  while (1) {
+    /* Check dbus */
+    FD_ZERO(&readfds);
+    FD_ZERO(&writefds);
+    FD_ZERO(&exceptfds);
+    tv.tv_sec = 0;
+    tv.tv_usec = 1000;
+    nfds = xcdbus_pre_select(g_xcbus, 0, &readfds, &writefds, &exceptfds);
+    select(nfds, &readfds, &writefds, &exceptfds, &tv);
```
Comment (rossphilipson) [inline src/main.c]: select() can fail - it should be checked - perhaps this code block is not done yet

File: src/main.c [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+    xcdbus_post_select(g_xcbus, 0, &readfds, &writefds, &exceptfds);
+    /* TODO: watch xenstore for new (domains and) vbds */
+  }
+
+  return ret;
```
Comment (rossphilipson) [inline src/main.c]: ret not used

File: src/blktap.c [hunk 10]
```diff
@@ -0,0 +1,345 @@
...
+  tmp = list;
+  while(*tmp != NULL) {
+    if ((*tmp)->minor == tap_minor) {
+      tap = *tmp;
+      break;
+    }
+    tmp++;
+  }
+  if (tap == NULL)
+    return false;
```
Comment (rossphilipson) [inline src/blktap.c:251]: Missing a tap_ctl_free_list(list); here?

**Comment by jean-edouard** [comment]
Pushed more fixes, thanks for the review.
We should probably wait until I get back from vacation on Nov 7 to test the fixes and merge.

**Comment by rossphilipson** [comment]
Ack we will leave this pending until you get back.

**Comment by jean-edouard** [comment]
The fixes seem to work, but I'm not fully satisfied with the logic yet...
Closing for now.

**Pull request closed without merging**